heron/api/src/java/com/twitter/heron/api/bolt/IStatefulBolt.java

```diff
@@ -0,0 +1,27 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.api.bolt;
+
+import com.twitter.heron.api.topology.IStatefulComponent;
+
+/**
+ * An IStatefulBolt represents a bolt that implements stateful functionality.
+ * In addition to the regular prepare/execute methods, it has initialization
+ * methods with state as defined in IStatefulComponent. An implementation
+ * of the stateful bolt saves state in the State interface and is able
+ * to restore its state from the State during init call.
+ */
+public interface IStatefulBolt extends IStatefulComponent, IRichBolt {
```

Comment:
IStatefulSpout/IStatefulBolt are mimicking IRichSpout/IRichBolt. They are essentially a shorthand version for any kind of stateful component.
IRich/IBasic stuff are essentially more developer friendly versions of the basic ISpout stuff. Thus in the future, depending on requirements, we might have IBasicStatefulBolt/Spout as well. They make it easy for developers to quickly start writing bolts/spouts.